File: scratch/todot.py
import json
from collections import Counter

filename = "seqcavlrecursivemapbasedcppcopy"

LEFT_COLOR = "crimson"
RITE_COLOR = "dodgerblue"
PARENT_COLOR = "darkseagreen2"
DEFAULT_NODE_COLOR = "white"
LEAF_COLOR = "forestgreen"
CHUNK_ROOT_COLOR = "gold1"
CHUNK_ROOT_AND_LEAF_COLOR = "palegreen2"


def get_obj():
    with open(f"dump/{filename}.json", "r") as f:
        return json.loads(f.read())


def pprint(o):
    pass


def dotnodelabel(node):
    id_ = node["id"]
    k = node["key"]
    v = node["value"]
    h = node["heightField"]
    label = f'"id: {id_} k: {k} v: {v} h: {h}"'
    return label


def dotnode(node):
    col = DEFAULT_NODE_COLOR
    if node["leaf"] and node["chunkRoot"]:
        col = CHUNK_ROOT_AND_LEAF_COLOR
    elif node["leaf"]:
        col = LEAF_COLOR
    elif node["chunkRoot"]:
        col = CHUNK_ROOT_COLOR
    return f"{dotnodelabel(node)} [style=filled, fillcolor={col}];"


def dotedge(src, dst, kind):
    col = None
    if kind == "left":
        col = LEFT_COLOR
    if kind == "rite":
        col = RITE_COLOR
    if kind == "parent":
        col = PARENT_COLOR

    src = dotnodelabel(src)
    dst = dotnodelabel(dst)
    return f"{src}->{dst} [style=bold, color={col}];"


def invisible_aux(parent):
    child_label = f'"invisible{parent["id"]}"'
    invisible_child = f"{child_label} [label={child_label}, style=invis];"
    invisible_edge = f"{dotnodelabel(parent)}->{child_label} [style=invis];"
    return invisible_child, invisible_edge


def graph(node_strings, edge_strings):
    header = """digraph{
ordering = out;
"""
    footer = """}
"""
    content = header
    for s in node_strings:
        content += s + "\n"
    for s in edge_strings:
        content += s + "\n"
    content += footer
    return content


edge_objs = get_obj()
node_objs = {}

nodes = {}
lefts = {}
rites = {}
parents = {}


for edge_obj in edge_objs:
    src = edge_obj["src"]
    dst = edge_obj["dst"]
    src_id = src["id"]
    dst_id = dst["id"]

    node_objs[src_id] = src
    node_objs[dst_id] = dst

    kind = edge_obj["type"]

    nodes[src_id] = dotnode(src)
    nodes[dst_id] = dotnode(dst)

    if kind == "left":
        lefts[src_id] = dotedge(src, dst, kind)
    if kind == "rite":
        rites[src_id] = dotedge(src, dst, kind)
    if kind == "parent":
        parents[src_id] = dotedge(src, dst, kind)

needs_left = set()
needs_rite = set()
for src_id, edge_str in lefts.items():
    if src_id not in rites.keys():
        needs_rite.add(src_id)
for src_id, edge_str in rites.items():
    if src_id not in lefts.keys():
        needs_left.add(src_id)

for node_id in needs_left:
    child, edge = invisible_aux(node_objs[node_id])
    nodes[-node_id] = child
    lefts[-node_id] = edge

for node_id in needs_rite:
    child, edge = invisible_aux(node_objs[node_id])
    nodes[-node_id] = child
    rites[-node_id] = edge

edges = list(lefts.values()) + list(rites.values()) + list(parents.values())
nodes = list(nodes.values())

print("\n".join(edges))
print("\n".join(nodes))

g = graph(nodes, edges)
with open(f"dump/{filename}.dot", "w") as f:
    f.write(g)

File: debugger/src/util/toDot.ts
import _ from "lodash";

enum EdgeType {
  LEFT = 1,
  RITE = 2,
  PARENT = 3,
}

type Edge = [string, string, string, EdgeType]; // [from, to, color, isLeft]

const LEFT_COLOR = "crimson";
const RITE_COLOR = "dodgerblue";
const PARENT_COLOR = "darkseagreen2";

const DEFAULT_NODE_COLOR = "white";
const LEAF_COLOR = "aquamarine";
const CHUNK_ROOT_COLOR = "gold1";
const CHUNK_ROOT_AND_LEAF_COLOR = "darkturquoise";

class Graph {
  keys: Record<string, string> = {};
  vals: Record<string, string> = {};
  heights: Record<string, number> = {};
  isChunkRoot: Record<string, boolean> = {};
  isLeaf: Record<string, boolean> = {};

  hasLeft: string[] = [];
  hasRite: string[] = [];

  edges: Edge[] = [];

  constructor() {
    // Label root
    this.keys["0"] = "RootHolder";
  }

  addLeftEdge(from: string, to: string) {
    this.hasLeft.push(from);
    this.edges.push([from, to, LEFT_COLOR, EdgeType.LEFT]);
  }

  addRiteEdge(from: string, to: string) {
    this.hasRite.push(from);
    this.edges.push([from, to, RITE_COLOR, EdgeType.RITE]);
  }

  addParentEdge(from: string, to: string) {
    this.edges.push([from, to, PARENT_COLOR, EdgeType.PARENT]);
  }

  addKey(address: string, k: string) {
    this.keys[address] = k;
  }

  addVal(address: string, v: string) {
    this.vals[address] = v;
  }

  addHeight(address: string, v: number) {
    this.heights[address] = v;
  }

  markChunkRoot(address: string) {
    this.isChunkRoot[address] = true;
  }

  markLeaf(address: string) {
    this.isLeaf[address] = true;
  }

  nodeLabel(address: string): string {
    var k = address in this.keys ? this.keys[address] : "None";
    var v = address in this.vals ? this.vals[address] : "None";
    var h = address in this.heights ? this.heights[address] : "?";
    return `"k:${k} v:${v} a:${address} h:${h}"`;
  }

  nodeStyle(address: string): string {
    var col = DEFAULT_NODE_COLOR;
    if (this.isChunkRoot[address] === true && this.isLeaf[address] === true) {
      col = CHUNK_ROOT_AND_LEAF_COLOR;
    } else if (this.isChunkRoot[address] === true) {
      col = CHUNK_ROOT_COLOR;
    } else if (this.isLeaf[address] === true) {
      col = LEAF_COLOR;
    }
    return `${this.nodeLabel(address)} [style=filled, fillcolor=${col}];`;
  }

  nodeStyleString(): string {
    var ret = "";
    for (const address of Object.keys(this.keys)) {
      ret += `${this.nodeStyle(address)}\n`;
    }
    return ret;
  }

  regularEdgeString(edge: Edge) {
    const [from, to, col, isLeft] = edge;
    return `${this.nodeLabel(from)} -> ${this.nodeLabel(
      to
    )} [style=bold, color=${col}];\n`;
  }

  edgeStrings(): string[] {
    var ret: string[] = [];

    _.filter(this.edges, (it) => it[3] === EdgeType.LEFT).forEach((it) => {
      ret.push(this.regularEdgeString(it));
    });

    ret = ret.concat(
      this.invisibleEdges(_.difference(this.hasRite, this.hasLeft))
    );

    _.filter(this.edges, (it) => it[3] === EdgeType.RITE).forEach((it) => {
      ret.push(this.regularEdgeString(it));
    });

    ret = ret.concat(
      this.invisibleEdges(_.difference(this.hasLeft, this.hasRite))
    );

    _.filter(this.edges, (it) => it[3] === EdgeType.PARENT).forEach((it) => {
      ret.push(this.regularEdgeString(it));
    });

    return ret;
  }

  invisibleEdgeString(addr: string): string {
    const label = `"Invis${addr}"`;
    return `
${label} [ label = ${label}, style = invis ]; ${this.nodeLabel(
      addr
    )} -> ${label} [ style = invis ];\n`.trimStart();
  }

  invisibleEdges(keys: string[]) {
    return keys.map((it) => this.invisibleEdgeString(it));
  }
}

export function wrapContent(content: string): string {
  function header() {
    return `
digraph {
ordering=out;
`;
  }

  function footer() {
    return `}`;
  }

  return header() + content + footer();
}

export default function toDot(data: any): string {
  // console.log("data", data);

  const left = data.left || {};
  const rite = data.rite || {};

  const key = data.key || {};
  const val = data.val || {};

  const height = data.height || {};

  const parent = data.parent || {};

  const nodeChunk = data.chunk_of_node || {};
  const isLeaf = data.is_leaf || {};

  const graph = new Graph();

  let content = "";

  for (const [k, v] of Object.entries(left)) {
    graph.addLeftEdge(k, (v as number).toString());
  }

  for (const [k, v] of Object.entries(rite)) {
    graph.addRiteEdge(k, (v as number).toString());
  }

  for (const [k, v] of Object.entries(parent)) {
    graph.addParentEdge(k, (v as number).toString());
  }

  for (const [k, v] of Object.entries(key)) {
    graph.addKey(k, (v as number).toString());
  }

  for (const [k, v] of Object.entries(val)) {
    graph.addVal(k, (v as number).toString());
  }

  for (const [k, v] of Object.entries(height)) {
    graph.addHeight(k, v as number);
  }

  for (const [k, v] of Object.entries(nodeChunk)) {
    if (v != null) {
      graph.markChunkRoot(k);
    }
  }

  for (const [k, v] of Object.entries(isLeaf)) {
    if (v != null) {
      graph.markLeaf(k);
    }
  }

  content += graph.nodeStyleString();

  graph.edgeStrings().forEach((it) => {
    content += it;
  });

  let ret = wrapContent(content);
  return ret;
}
